AccessControl Provider (#1229)
pankod/refine: documentation/docs/guides-and-concepts/access-control.md

---
id: access-control
title: Access Control
sidebar_label: Access Control
---

import simpleAccess from '@site/static/img/guides-and-concepts/access-control/simple-access.png';

## Introduction

> Access control is a broad topic where there are lots of advanced solutions that provide different set of features. **refine** is deliberately agnostic for its own API to be able to integrate different solutions(RBAC, ABAC, ACL, [casbin](https://casbin.org/), [casl](https://casl.js.org/v5/en/), [cerbos](https://cerbos.dev/), [accesscontrol.js](https://onury.io/accesscontrol/)). `can` method would be the entry point for those solutions.

Reviewer: ```suggestion
Access control is a broad topic where there are lots of advanced solutions that provide different set of features. **refine** is deliberately agnostic for its own API to be able to integrate different solutions(RBAC, ABAC, ACL, IP, LDAP, etc...). `can` method would be the entry point for those solutions.
```

[Refer to the Access Control Provider documentation for detailed information. &#8594](api-references/providers/accessControl-provider.md)

**refine** provides an agnostic API via the `accessControlProvider` to manage access control throughout your app.

An `accessControlProvider` must implement only one async method named `can` to be used to check if the desired access will be granted.

> We will be using **casbin** in this guide for users with different roles who have different access rights for parts of the app.

Reviewer: ```suggestion
We will be using **[Casbin](https://casbin.org/)** in this guide for users with different roles who have different access rights for parts of the app.
```

## Installation

> We need to install casbin.

Reviewer: ```suggestion
We need to install Casbin.
```

```bash
npm install casbin.js --save
```

## Setup

> The app will have three resources: **posts**, **users** and **categories** with crud pages(list, create, edit and show)

Reviewer: ```suggestion
The app will have three resources: **posts**, **users** and **categories** with CRUD pages(list, create, edit and show)
```

import {
    CategoryList,
    CategoryCreate,
    CategoryEdit,
    CategoryShow,
} from "pages/categories";

const API_URL = "https://api.fake-rest.refine.dev";

const App: React.FC = () => {
    return (
        <Refine
            routerProvider={routerProvider}
            dataProvider={dataProvider(API_URL)}
            resources={[
                {
                    name: "posts",
                    list: PostList,
                    create: PostCreate,
                    edit: PostEdit,
                    show: PostShow,
                    canDelete: true,
                },
                {
                    name: "users",
                    list: UserList,
                    create: UserCreate,
                    edit: UserEdit,
                    show: UserShow,
                },
                {
                    name: "categories",
                    list: CategoryList,
                    create: CategoryCreate,
                    edit: CategoryEdit,
                    show: CategoryShow,
                },
            ]}
        />
    );
};

export default App;

```

## Adding Policy and Model

> The way **casbin** works is that access rights are checked according to policies that are defined based on a model. [You can find further information about how models and policies work here](https://casbin.org/docs/en/how-it-works).

Reviewer: ```suggestion
The way **[Casbin](https://casbin.org/)** works is that access rights are checked according to policies that are defined based on a model. [You can find further information about how models and policies work here](https://casbin.org/docs/en/how-it-works).
```

```ts title="src/accessControl.ts"
import { newModel, MemoryAdapter } from "casbin.js";

export const model = newModel(`
[request_definition]
r = sub, obj, act

[policy_definition]
p = sub, obj, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && keyMatch(r.obj, p.obj) && regexMatch(r.act, p.act)
`);

export const adapter = new MemoryAdapter(`
p, editor, posts, (list)
`);

```

:::tip
You can can find more examples in [casbin documentation](https://casbin.org/docs/en/supported-models) or play with lots of examples in [casbin editor](https://casbin.org/en/editor)  
:::

## Adding `accessControlProvider`

Now we will implement the `can` method for `accessControlProvider` to integrate our policy.

```tsx title="src/App.tsx" 
// ...
import { newEnforcer } from "casbin.js";

import { model, adapter } from "accessControl";

const App: React.FC = () => {
    return (
        <Refine
            // ...
            // highlight-start
            accessControlProvider={{
                can: async ({ resource, action, params }) => {

>                     const enforcer = await newEnforcer(model, adapter);
>                     
>                     return Promise.resolve({
>                         can: await enforcer.enforce("editor", resource, action),
>                     });

Reviewer: ```suggestion
                    const enforcer = await newEnforcer(model, adapter);
                    const can = await enforcer.enforce("editor", resource, action);

                    return Promise.resolve({
                        can,
                    });
```

import dynamicRole from '@site/static/img/guides-and-concepts/access-control/dynamic-role.gif';

## Introduction

Access control is a broad topic where there are lots of advanced solutions that provide different set of features. **refine** is deliberately agnostic for its own API to be able to integrate different solutions(RBAC, ABAC, ACL, IP, LDAP, etc...). `can` method would be the entry point for those solutions.

[Refer to the Access Control Provider documentation for detailed information. &#8594](api-references/providers/accessControl-provider.md)

**refine** provides an agnostic API via the `accessControlProvider` to manage access control throughout your app.

An `accessControlProvider` must implement only one async method named `can` to be used to check if the desired access will be granted.

We will be using **[Casbin](https://casbin.org/)** in this guide for users with different roles who have different access rights for parts of the app.

## Installation

We need to install Casbin.

```bash
npm install casbin.js --save
```

## Setup

The app will have three resources: **posts**, **users** and **categories** with CRUD pages(list, create, edit and show)

> [You can refer to codesandbox to how they are implemented](#live-codesandbox-example)

Reviewer 2: ```suggestion
[You can refer to codesandbox to see how they are implemented](#live-codesandbox-example)
```

[request_definition]
r = sub, obj, act

[policy_definition]
p = sub, obj, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && keyMatch(r.obj, p.obj) && regexMatch(r.act, p.act)
`);

export const adapter = new MemoryAdapter(`
p, editor, posts, list
`);

```

:::tip
You can can find more examples in [casbin documentation](https://casbin.org/docs/en/supported-models) or play with lots of examples in [casbin editor](https://casbin.org/en/editor)  
:::

## Adding `accessControlProvider`

Now we will implement the `can` method for `accessControlProvider` to integrate our policy.

```tsx title="src/App.tsx" 
// ...
// highlight-next-line
import { newEnforcer } from "casbin.js";

// highlight-next-line
import { model, adapter } from "accessControl";

const App: React.FC = () => {
    return (
        <Refine
            // ...
            // highlight-start
            accessControlProvider={{
                can: async ({ resource, action }) => {
                    const enforcer = await newEnforcer(model, adapter);
                    const can = await enforcer.enforce("editor", resource, action);

>                     return Promise.resolve({ can });

Reviewer 2: ```suggestion
                    return { can };
```

p, admin, posts, (list)|(create)
p, admin, users, (list)|(create)
p, admin, categories, (list)|(create)

p, editor, posts, (list)|(create)
p, editor, categories, list
`);
```

- **admin** will have access to **list** and **create** for every resource
- **editor** will have access to **list** and **create** for **posts**
- **editor** won't have any access for **users**
- **editor** will have only **list** access for **categories**

We can demonstrate the effect of different roles by changing the `role` dynamically. Let's implement a switch in the header for selecting either **admin** or **editor** role to see the effect on the app.

```tsx title="src/App.tsx" 
// ...
// highlight-next-line
import { Header } from "components/header";

const App: React.FC = () => {
    // highlight-next-line
    const [role, setRole] = useState("admin");

    return (
        <Refine
            // ...
            accessControlProvider={{
                can: async ({ resource, action }) => {
                    const enforcer = await newEnforcer(model, adapter);
                    // highlight-next-line
                    const can = await enforcer.enforce(role, resource, action);

>                     return Promise.resolve({
>                         can,
>                     });

Reviewer 2: ```suggestion
                    return {
                        can,
                    };
```

`);
```

- **admin** will have **edit**, **show** and **delete** access for every resource
- **editor** will have **edit** and **show** access for **posts**

:::tip
`*` is wildcard. Specific ids can be targeted too. For example If you want **editor** role to have **delete** access for **post** with **id** `5`, you can add this policy:

```ts
export const adapter = new MemoryAdapter(`
p, editor, posts/5, delete
`);
```
:::

We must handle id based access controls in the `can` method. **id** parameter will be accessible in `params`.

```tsx title="src/App.tsx"
// ...

const App: React.FC = () => {
    // ...

    return (
        <Refine
            // ...
            accessControlProvider={{
                // highlight-start
                can: async ({ resource, action, params }) => {
                    const enforcer = await newEnforcer(model, adapter);
                    if ( action === "delete" || action === "edit" || action === "show" ) {
                        const can = await enforcer.enforce(
                            role,
                            `${resource}/${params.id}`,
                            action,
                        )

>                         return Promise.resolve({ can });

Reviewer 2: ```suggestion
                        return { can };
```

                    }
                    // highlight-end
                    
                    const can = await enforcer.enforce(role, resource, action);

>                     return Promise.resolve({ can });

Reviewer 2: ```suggestion
                    return { can };
```
